internal/controller/helmrelease_controller.go

```diff
@@ -717,36 +770,144 @@ func (r *HelmReleaseReconciler) requestsForHelmChartChange(ctx context.Context,
-		// temporary errors which do not have an impact as long as
-		// there is an Artifact for the current Generation.
 		if conditions.IsFalse(obj, meta.ReadyCondition) {
 			msg = conditions.GetMessage(obj, meta.ReadyCondition)
 		}
-		return false, msg
+		return false, fmt.Sprintf("%s '%s/%s' is not ready: %s",
+			kind, obj.GetNamespace(), obj.GetName(), msg)
 	case conditions.IsStalled(obj):
-		return false, conditions.GetMessage(obj, meta.StalledCondition)
-	case obj.Status.Artifact == nil:
-		return false, "does not have an artifact"
+		return false, fmt.Sprintf("%s '%s/%s' is not ready: %s",
+			kind, obj.GetNamespace(), obj.GetName(), conditions.GetMessage(obj, meta.StalledCondition))
+	case artifact == nil:
+		return false, fmt.Sprintf("%s '%s/%s' is not ready: %s",
+			kind, obj.GetNamespace(), obj.GetName(), "does not have an artifact")
 	default:
 		return true, ""
 	}
 }
+
+func isValidChartRef(obj *v2.HelmRelease) bool {
+	return (obj.HasChartRef() && !obj.HasChartTemplate()) ||
+		(!obj.HasChartRef() && obj.HasChartTemplate())
+}
+
+func getNamespacedName(obj *v2.HelmRelease) (types.NamespacedName, error) {
+	namespacedName := types.NamespacedName{}
+	switch {
+	case obj.HasChartRef() && !obj.HasChartTemplate():
+		namespacedName.Namespace = obj.Spec.ChartRef.Namespace
+		namespacedName.Name = obj.Spec.ChartRef.Name
+	case !obj.HasChartRef() && obj.HasChartTemplate():
+		namespacedName.Namespace = obj.Spec.Chart.GetNamespace(obj.GetNamespace())
+		namespacedName.Name = obj.GetHelmChartName()
+	default:
+		return namespacedName, fmt.Errorf("one of chartRef or chart must be present")
+	}
+
+	return namespacedName, nil
+}
+
+func mutateChartWithSourceRevision(chart *chart.Chart, source source.Source) (string, error) {
+	// If the source is an OCIRepository, we can try to mutate the chart version
+	// with the artifact revision. The revision is either a <tag>@<digest> or
+	// just a digest.
+	obj, ok := source.(*sourcev1.OCIRepository)
+	if !ok {
+		// if not make sure to return an empty string to delete the digest of the
+		// last attempted revision
+		return "", nil
+	}
+	ver, err := semver.NewVersion(chart.Metadata.Version)
+	if err != nil {
+		return "", err
+	}
+
+	var ociDigest string
+	revision := obj.GetArtifact().Revision
+	switch {
+	case strings.Contains(revision, "@"):
+		tagD := strings.Split(revision, "@")
+		if len(tagD) != 2 || tagD[0] != chart.Metadata.Version {
+			return "", fmt.Errorf("artifact revision %s does not match chart version %s", tagD[0], chart.Metadata.Version)
+		}
+		// algotithm are sha256, sha384, sha512 with the canonical being sha256
+		// So every digest starts with a sha algorithm and a colon
+		sha := strings.Split(tagD[1], ":")[1]
+		// add the digest to the chart version to make sure mutable tags are detected
+		*ver, err = ver.SetMetadata(sha[0:12])
+		if err != nil {
+			return "", err
+		}
+		ociDigest = tagD[1]
+	default:
+		// default to the digest
+		sha := strings.Split(revision, ":")[1]
```

Comment:
We should check the split result and error out instead of causing a panic if there is no `:` in the revision string.